Sources/Plasma/FeatureLib/pfGLPipeline/plGLEnumerate.cpp

```diff
@@ -0,0 +1,306 @@
+        pfd.cDepthBits = 24;
+        pfd.iLayerType = PFD_MAIN_PLANE;
+
+        int format = ChoosePixelFormat(dc, &pfd);
+        if (!format)
+            break;
+
+        if (!SetPixelFormat(dc, format, &pfd))
+            break;
+
+        ctx = wglCreateContext(dc);
+        if (!ctx)
+            break;
+
+        if (!wglMakeCurrent(dc, ctx))
+            break;
+
+        hsG3DDeviceRecord devRec;
+        devRec.SetG3DDeviceType(hsG3DDeviceSelector::kDevTypeOpenGL);
+        devRec.SetDriverName("opengl32.dll");
+
+        if (fillDeviceRecord(devRec))
+            records.emplace_back(devRec);
+    } while (0);
+
+    // Cleanup:
+    if (ctx) {
+        wglMakeCurrent(nullptr, nullptr);
+        wglDeleteContext(ctx);
+    }
+
+    if (dc)
+        ReleaseDC(wnd, dc);
+
+    if (wnd)
+        DestroyWindow(wnd);
+
+    if (cls)
+        UnregisterClass(reinterpret_cast<LPCTSTR>(cls), GetModuleHandle(nullptr));
+}
+#endif // HS_BUILD_FOR_WIN32
+
+
+#ifdef HS_BUILD_FOR_MACOS
+#include <OpenGL/OpenGL.h>
+
+void plCGLEnumerate(std::vector<hsG3DDeviceRecord>& records)
+{
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    CGLPixelFormatObj pix = nullptr;
+    CGLContextObj ctx = nullptr;
+
+    do {
+        CGLPixelFormatAttribute attribs[6] = {
+            kCGLPFAAccelerated,
+            kCGLPFANoRecovery,
+            kCGLPFADoubleBuffer,
+            kCGLPFAOpenGLProfile, (CGLPixelFormatAttribute) kCGLOGLPVersion_3_2_Core,
+            (CGLPixelFormatAttribute) 0
+        };
+
+        int nPix = 0;
+        if (CGLChoosePixelFormat(attribs, &pix, &nPix) != kCGLNoError || nPix == 0)
+            break;
+
+        if (CGLCreateContext(pix, nullptr, &ctx) != kCGLNoError)
+            break;
+
+        CGLSetCurrentContext(ctx);
+
+        hsG3DDeviceRecord devRec;
+        devRec.SetG3DDeviceType(hsG3DDeviceSelector::kDevTypeOpenGL);
+        devRec.SetDriverName("OpenGL.framework");
+
+        if (fillDeviceRecord(devRec))
+            records.emplace_back(devRec);
+    } while (0);
+
+    // Cleanup:
+    if (ctx) {
+        CGLSetCurrentContext(nullptr);
+        CGLReleaseContext(ctx);
+    }
+
+    if (pix)
+        CGLReleasePixelFormat(pix);
+#pragma clang diagnostic pop
+}
+#endif // HS_BUILD_FOR_MACOS
+
+void plGLEnumerate(std::vector<hsG3DDeviceRecord>& records)
+{
+#ifdef USE_EGL
+    // The USE_EGL define tells us whether the epoxy library includes support
+    // for attempting to use EGL on the current platform, but we still need to
+    // check if EGL is actually available at runtime.
+    //
+    // On Windows, this may be true in cases like the PowerVR SDK or when using
+    // ANGLE.
+    //
+    // On Linux, this should be true with mesa or nvidia drivers.
+    if (epoxy_has_egl()) {
+        plEGLEnumerate(records);
+    }
+#endif
+
+#ifdef HS_BUILD_FOR_WIN32
+    plWGLEnumerate(records);
+#endif
+
+#ifdef HS_BUILD_FOR_MACOS
+    plCGLEnumerate(records);
+#endif
```

Comment:
Should there be handling for a not supported platform?

Reply:
No, in that case there will be no DeviceRecord added to the records vector, and hsG3DDeviceSelector will report that OpenGL is not available (and either fail over to something else, or (more likely, since GL is always going to be the second choice when other options are supported) cause plClient to show an error about no appropriate rendering devices)